### PR TITLE
Add Java port

### PR DESCRIPTION
**Pull request by pradyunsg**
Part of #23

Any chance of someone running :eyes: through this?

### CODE REVIEW

**Comment by getify** [comment]
Is it appropriate for the tests to be in the same file as the implementation? This seems strange to me.

**Comment by pradyunsg** [comment]
I agree. I'm not very fluent in Java (and I'd presume neither are you). So, any ideas?

**Comment by pradyunsg** [comment]
Also, shouldn't the README.md be modified?

**Comment by getify** [comment]
> shouldn't the README.md be modified?

Yes, see the modifications I made to the `README.md` in the "javascript", "node" and "php" branches as minimal required updates for each port.

**Comment by pradyunsg** [comment]
But what should the modifications be?

**Comment by getify** [comment]
As indicated in the PORTING.md, item (4), I'd prefer to see something that's consistent with each language's basic conventions. In general, I believe most languages keep implementation and tests separate. I can't think of a strong reason to combine those separate concerns in a single file for any language.

**Comment by pradyunsg** [comment]
I think I can split the Java file. I'll get to it later today or tomorrow.
Just that, what should the package name be? `com.getify.json_minify` is what I'm thinking...

PS: I should split the python tests as well. Right?

**Comment by getify** [comment]
Can we do `com.json_minify`? I dunno enough about Java to know if that's good or bad, but I don't see why this needs the `getify` moniker in it. I don't have any other Java software that anyone would need to namespace in.

> I should split the python tests as well. Right?

Yes please. :)

**Comment by pradyunsg** [comment]
I guess we can.

Added to tomorrow's TODO list.

**Comment by kag0** [comment]
If this is still unresolved... If the java repo is created, I can get this set up with maven and move the tests into appropriate locations with junit. If there is further interest we can get things set up on maven central or in jipack.

**Comment by pradyunsg** [comment]
@kag0 

> If this is still unresolved...

It is indeed unresolved. 

> If the java repo is created, 

I'm assuming you mean the [`java`][java-branch-link] branch on this repository.

> I can get this set up with maven and move the tests into appropriate locations with junit.

Please do. I wanted to do this but don't really have that much knowledge (and now, time to gain it) to do this.
> If there is further interest we can get things set up on maven central or in jipack.

I'm interested. :smile:

------
Please do read PORTING.md and TESTING.md files. You'll should to make a PR to the `java` branch in this repository. You could include the work I've done (this PR) or start afresh off the `java` branch in the repository.

[java-branch-link]: https://github.com/getify/JSON.minify/tree/java

**Comment by kag0** [comment]
Sounds good. Let's merge this PR for now and open issues regarding creating an account and signing keys for maven central, and maybe one for CI/CD.

**Comment by pradyunsg** [comment]
> Sounds good. Let's merge this PR for now 

Done.

> open issues regarding creating an account and signing keys for maven central, and maybe one for CI/CD.

Go ahead and open those issues.

> also, do you mind if I add some inline comments to this PR for improvements? (I'll do them all when I separate the tests if they aren't done before merging the PR)

No I don't. :smile:

File: Minify.java [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.junit.Test;
+
+public class Minify {
+
+    public String minify(String jsonString) {
+        String tokenizer = "\"|(/\\*)|(\\*/)|(//)|\\n|\\r";
```
Comment (kag0) [inline Minify.java]: Make these strings private static final class variables.

File: Minify.java [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+        Boolean in_multiline_comment = false;
+        Boolean in_singleline_comment = false;
+        String tmp = "";
+        String tmp2 = "";
+        List<String> new_str = new ArrayList<String>();
+        Integer from = 0;
+        String lc = "";
+        String rc = "";
+
+        Pattern pattern = Pattern.compile(tokenizer);
```
Comment (kag0) [inline Minify.java]: Patterns are thread safe, make this PSF.

File: Minify.java [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+    public String minify(String jsonString) {
+        String tokenizer = "\"|(/\\*)|(\\*/)|(//)|\\n|\\r";
+        String magic = "(\\\\)*$";
+        Boolean in_string = false;
+        Boolean in_multiline_comment = false;
+        Boolean in_singleline_comment = false;
+        String tmp = "";
+        String tmp2 = "";
+        List<String> new_str = new ArrayList<String>();
```
Comment (kag0) [inline Minify.java]: This should be a StringBuilder

File: Minify.java [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+                "in case you are interested. */\n" +
+                "\"something\": 10,\n" +
+                "\"else\": 20\n" +
+                "}\n" +
+                "/* NOTE: You can easily strip the whitespace and comments\n" +
+                "   from such a file with the JSON.minify() project hosted \n"+
+                "   here on github at http://github.com/getify/JSON.minify \n"+
+                "*/";
+
+        String test1_res = "{\"foo\":\"bar\",\"bar\":[\"baz\",\"bum\",\"zam\"],\"something\":10,\"else\":20}";
```
Comment (kag0) [inline Minify.java]: split the tests out into their own methods

File: Minify.java [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+        assert(minify(test1).equals(test1_res));
+
+        System.out.println("Running test2...");
+        String test2 = "{\"/*\":\"*/\",\"//\":\"\",/*\"//\"*/\"/*/\"://\n" +
+                      "\"//\"}" +
+                      "";
+
+        String test2_res = "{\"/*\":\"*/\",\"//\":\"\",\"/*/\":\"//\"}";
+        assert(minify(test2).equals(test2_res));
```
Comment (kag0) [inline Minify.java]: use assertEquals

File: Minify.java [hunk 6]
```diff
@@ -0,0 +1,164 @@
+//
+//  Minify.java
+//  jsonminifier - Ported from https://github.com/getify/JSON.minify
+//
+//  Created by Bernhard Gass on 8/01/13.
+//  Copyright © 2013 Bernhard Gass. All rights reserved.
+
+package com.changeme.json;
```
Comment (kag0) [inline Minify.java]: change it

File: Minify.java [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.junit.Test;
+
+public class Minify {
```
Comment (kag0) [inline Minify.java]: make class abstract. don't want anyone instantiating this

**Pull request merged**